bootstrap/service.go

```diff
@@ -0,0 +1,308 @@
+
+// Service specifies an API that must be fulfilled by the domain service
+// implementation, and all of its decorators (e.g. logging & metrics).
+type Service interface {
+	// Add adds new Thing to the user identified by the provided key.
+	Add(string, Config) (Config, error)
+
+	// View returns Thing with given ID belonging to the user identified by the given key.
+	View(string, string) (Config, error)
+
+	// Update updates editable fields of the provided Thing.
+	Update(string, Config) error
+
+	// List returns subset of Things with given state that belong to the user identified by the given key.
+	List(string, Filter, uint64, uint64) ([]Config, error)
+
+	// Remove removes Thing with specified key that belongs to the user identified by the given key.
+	Remove(string, string) error
+
+	// Bootstrap returns configuration to the Thing with provided external ID using external key.
+	Bootstrap(string, string) (Config, error)
+
+	// ChangeState changes state of the Thing with given ID and owner.
+	ChangeState(string, string, State) error
+}
+
+// ConfigReader is used to parse Config into format which will be encoded
+// as a JSON and consumed from the client side. The purpose of this interface
+// is to provide convenient way to generate custom configuration response
+// based on the specific Config which will be consumed by the client.
+type ConfigReader interface {
+	ReadConfig(Config) (mainflux.Response, error)
+}
+
+type bootstrapService struct {
+	users  mainflux.UsersServiceClient
+	things ConfigRepository
+	sdk    mfsdk.SDK
+}
+
+// New returns new Bootstrap service.
+func New(users mainflux.UsersServiceClient, things ConfigRepository, sdk mfsdk.SDK) Service {
+	return &bootstrapService{
+		things: things,
+		sdk:    sdk,
+		users:  users,
+	}
+}
+
+func (bs bootstrapService) Add(key string, thing Config) (Config, error) {
+	owner, err := bs.identify(key)
+	if err != nil {
+		return Config{}, err
+	}
+	// Check if channels exist. This is the way to prevent invalid configuration to be saved.
+	// However, channels deletion wil eventually cause this; since Bootstrap service is not
+	// using events from the Things service at the moment.
+	for _, c := range thing.MFChannels {
+		if _, err := bs.sdk.Channel(c, key); err != nil {
+			return Config{}, ErrMalformedEntity
+		}
+	}
+	mfThing, err := bs.add(key)
+	if err != nil {
+		return Config{}, err
+	}
+
+	thing.MFThing = mfThing.ID
+	thing.Owner = owner
+	thing.State = Inactive
+	thing.MFKey = mfThing.Key
+
+	id, err := bs.things.Save(thing)
+	if err != nil {
+		return Config{}, err
+	}
+	bs.things.RemoveUnknown(thing.ExternalKey, thing.ExternalID)
+
+	thing.MFThing = id
+	return thing, nil
+}
+
+func (bs bootstrapService) View(key, id string) (Config, error) {
+	owner, err := bs.identify(key)
+	if err != nil {
+		return Config{}, err
+	}
+	return bs.things.RetrieveByID(owner, id)
+}
+
+func (bs bootstrapService) Update(key string, thing Config) error {
```

Comment:
This function is really long, can you break it into multiple private functions (if there are more than one logical step)?